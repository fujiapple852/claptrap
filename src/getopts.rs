use crate::arg::{Arg, ArgAction};
use crate::command::Command;
use crate::num_args::NumArgs;

/// Parse a classic getopts specification string and build a [`Command`].
///
/// The returned `Command` will have a default name of `"prog"` and one
/// argument per option found in the getopts string.
pub fn parse_getopts(spec: &str) -> anyhow::Result<Command> {
    let mut chars = spec.chars().peekable();
    if chars.peek() == Some(&':') {
        // skip leading ':' which toggles silent error mode in `getopts`
        chars.next();
    }

    let mut command = Command::new("prog");
    while let Some(ch) = chars.next() {
        if ch == ':' || ch.is_whitespace() {
            return Err(anyhow::anyhow!("invalid option character: {ch}"));
        }

        let mut arg = Arg::new().short(ch);
        let mut action = ArgAction::SetTrue;

        let mut colon_count = 0;
        while matches!(chars.peek(), Some(':')) {
            colon_count += 1;
            chars.next();
        }

        match colon_count {
            0 => {}
            1 => {
                action = ArgAction::Set;
                arg = arg.num_args(NumArgs::Exact(1));
            }
            2 => {
                action = ArgAction::Set;
                arg = arg.num_args(NumArgs::RangeToInclusive(..=1));
            }
            _ => {
                return Err(anyhow::anyhow!(
                    "too many ':' after option {ch}"
                ));
            }
        }

        arg = arg.action(action);
        command = command.arg(ch.to_string(), arg);
    }

    Ok(command)
}
