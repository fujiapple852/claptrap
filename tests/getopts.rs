use claptrap::getopts::parse_getopts;

#[test]
fn parse_empty() {
    let cmd = parse_getopts("").unwrap();
    insta::assert_debug_snapshot!(cmd);
}

#[test]
fn parse_leading_colon_only() {
    let cmd = parse_getopts(":").unwrap();
    insta::assert_debug_snapshot!(cmd);
}

#[test]
fn parse_basic_spec() {
    let cmd = parse_getopts("ab:c::").unwrap();
    insta::assert_debug_snapshot!(cmd);
}

#[test]
fn error_triple_colon() {
    let err = parse_getopts("a:::").unwrap_err();
    insta::assert_snapshot!(err.to_string());
}

#[test]
fn error_stray_colon() {
    let err = parse_getopts("::").unwrap_err();
    insta::assert_snapshot!(err.to_string());
}

#[test]
fn error_whitespace() {
    let err = parse_getopts("a b").unwrap_err();
    insta::assert_snapshot!(err.to_string());
}
