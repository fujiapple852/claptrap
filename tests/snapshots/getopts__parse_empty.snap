---
source: tests/getopts.rs
expression: cmd
---
Command {
    name: "prog",
    args: None,
    subcommands: None,
    groups: None,
    ignore_errors: None,
    args_override_self: None,
    dont_delimit_trailing_values: None,
    color: None,
    term_width: None,
    max_term_width: None,
    disable_version_flag: None,
    propagate_version: None,
    next_line_help: None,
    disable_help_flag: None,
    disable_help_subcommand: None,
    disable_colored_help: None,
    help_expected: None,
    hide_possible_values: None,
    infer_long_args: None,
    infer_subcommands: None,
    bin_name: None,
    display_name: None,
    author: None,
    about: None,
    long_about: None,
    after_help: None,
    after_long_help: None,
    before_help: None,
    before_long_help: None,
    version: None,
    long_version: None,
    override_usage: None,
    override_help: None,
    help_template: None,
    flatten_help: None,
    next_help_heading: None,
    next_display_order: None,
    arg_required_else_help: None,
    allow_missing_positional: None,
    short_flag: None,
    long_flag: None,
    alias: None,
    short_flag_alias: None,
    long_flag_alias: None,
    aliases: None,
    short_flag_aliases: None,
    long_flag_aliases: None,
    visible_alias: None,
    visible_short_flag_alias: None,
    visible_long_flag_alias: None,
    visible_aliases: None,
    visible_short_flag_aliases: None,
    visible_long_flag_aliases: None,
    display_order: None,
    hide: None,
    subcommand_required: None,
    allow_external_subcommands: None,
    args_conflicts_with_subcommands: None,
    subcommand_precedence_over_arg: None,
    subcommand_negates_reqs: None,
    multicall: None,
    subcommand_value_name: None,
    subcommand_help_heading: None,
}
